### PR TITLE
Ritam auth2

### LINKED ISSUE #12: Authentication
#### As an application hoster I want to add user authentication to my application.

__Story__:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

Please split this task up into three separate PRs. The second, and third should
depend on the preceding ones.

__Acceptance criteria PR1__:
- [x] Users can sign up.

__Acceptance criteria PR2__:
- [x] Users can log in after they've signed up.
- [x] Users can log out after logging in.
- [x] Logged in users should not be able to register, nor log in again.

__Acceptance criteria PR3__:
- [ ] Users can choose to be remembered as part of the login
  process.<sup>†</sup>
- [ ] Multiple browsers can concurrently maintain the remembered status for the
  same user.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser. Please be sure to completely
close the browser (`cmd` + `q`) when testing this functionality. Closing the
window (`cmd` + `w`, or clicking the red circle) on Mac OS does not close the
browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/modeling_users
- https://www.railstutorial.org/book/sign_up
- https://www.railstutorial.org/book/basic_login
- https://www.railstutorial.org/book/advanced_login

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by RitamBanerjee**
Closes #12.2

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  get    '/login',   to: 'sessions#new'
```
Comment (grahamotte) [inline config/routes.rb]: Why use these instead of the rails standard?

File: test/controllers/sessions_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+require 'test_helper'
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
+  def setup
+    @user = User.new(email: 'abc@gmail.com', password: 'password', password_confirmation: 'password')
```
Comment (grahamotte) [inline test/controllers/sessions_controller_test.rb]: What happens if this fails?

File: .rubocop.yml [hunk 3]
```diff
@@ -31,3 +31,6 @@ Style/MultilineMethodCallIndentation:
 
 Metrics/MethodLength:
   Enabled: false
+
+Metrics/AbcSize:
```
Comment (grahamotte) [inline .rubocop.yml]: Do we need this?

File: test/controllers/sessions_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+require 'test_helper'
+
+class SessionsControllerTest < ActionDispatch::IntegrationTest
```
Comment (grahamotte) [inline test/controllers/sessions_controller_test.rb]: I think your tests need to be more comprehensive here

File: test/flows/sessions_flow_test.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+require 'flow_test_helper'
+
+class SessionsFlowTest < FlowTestCase
+  test 'log in and log out' do
```
Comment (grahamotte) [inline test/flows/sessions_flow_test.rb]: You also made a change to your user creation that a user gets auto logged in, so that test as well should be updated as well

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: What happens if the user is already logged in?

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      log_in user
+      flash[:success] = 'You are logged in!'
+      redirect_to images_path
+    else
+      flash[:error] = 'User not found'
+      render :new, status: :unauthorized
+    end
+  end
+
+  def destroy
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: similarly, what if they are logged out already?

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: You should use strong params for this

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      flash[:success] = 'You are logged in!'
+      redirect_to images_path
+    else
+      flash[:error] = 'User not found'
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: Does this need to be a `flash.now`? Also are there other reasons this could error other than user not found?

File: test/controllers/sessions_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  def setup
+    @user = User.new(email: 'abc@gmail.com', password: 'password', password_confirmation: 'password')
+    @user.save!
+  end
+  test 'should get new' do
+    get login_path
+    assert_response :success
+  end
+  test 'should be able to log in' do
```
Comment (grahamotte) [inline test/controllers/sessions_controller_test.rb]: Should assert the flash and the session in these tests

File: test/flows/sessions_flow_test.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+    @user = User.new(email: 'abc@gmail.com', password: 'password', password_confirmation: 'password')
+    @user.save!
+
+    sessions_login_page = PageObjects::Sessions::NewPage.visit
+
+    logged_in_index_page = sessions_login_page.login_user!(email: 'abc@gmail.com',
+                                                           password: 'password')
+
+    assert_equal 'You are logged in!', logged_in_index_page.flash_message(:success).text
```
Comment (grahamotte) [inline test/flows/sessions_flow_test.rb]: Maybe the login should fail the first time so we can cover that branch as well

File: .rubocop.yml (hunk 3 above)
Comment (RitamBanerjee) [reply .rubocop.yml]: Without it Rubocop was complaining about the length of this if: if user && user.authenticate(params[:session][:password])

File: config/routes.rb (hunk 1 above)
Comment (RitamBanerjee) [reply config/routes.rb]: yeah I could have left it the standard, but the rails tutorial linked in the issue did it this way and i thought it made sense

File: .rubocop.yml (hunk 3 above)
Comment (grahamotte) [reply .rubocop.yml]: Perhaps we should refactor that instead

File: config/routes.rb (hunk 1 above)
Comment (grahamotte) [reply config/routes.rb]: IMO it makes these routes less maintainable because you
- have to know that `login_path` goes to `sessions#whatever`
- have to update those helpers if you ever change your routes

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        if !logged_in?
+          log_in user
+          flash[:success] = 'You are logged in!'
+          redirect_to(images_path) && return
+        else
+          flash.now[:error] = 'User already logged in'
+          render(:new, status: :unauthorized) && return
+        end
+      else
+        flash.now[:error] = 'Incorrect Password'
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: For this it's usually a good idea to just say something like 'incorrect login/pass combo' so that you don't expose who has an account

File: app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+      flash.now[:error] = 'Unable to find your account'
+      render(:new, status: :unauthorized) && return
+    end
+  end
+
+  def destroy
+    if current_user
+      log_out
+    else
+      flash[:error] = 'Not logged in!'
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: Seems like this already logged in/out branch could be extracted to a `before_action`

File: .rubocop.yml [hunk 14]
```diff
@@ -31,3 +31,12 @@ Style/MultilineMethodCallIndentation:
 
 Metrics/MethodLength:
   Enabled: false
+
+Metrics/PerceivedComplexity:
```
Comment (grahamotte) [inline .rubocop.yml]: I think at least the sake of ropes you should refrain from disabling rubocop validations

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+  private
+
+  def check_login
+    @user = User.find_by(email: permitted_session_params[:email].downcase)
+    if @user
+      if @user.authenticate(permitted_session_params[:password])
+
+      else
+        flash.now[:error] = 'Incorrect Password/Email Combination'
+        render(:new, status: :unauthorized) && return
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: note: usually `&&` is used for conditionals and `and` is used for control flow (here). However it doesn't seem bad practice to use `&&` all the time.

File: app/models/user.rb [hunk 16]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  include BCrypt
```
Comment (grahamotte) [inline app/models/user.rb]: Why did you add this?

File: db/schema.rb [hunk 17]
```diff
@@ -44,4 +44,11 @@
     t.index ["name"], name: "index_tags_on_name", unique: true
   end
 
+  create_table "users", force: :cascade do |t|
```
Comment (grahamotte) [inline db/schema.rb]: Wasn't this already added?

File: test/flows/sessions_flow_test.rb [hunk 18]
```diff
@@ -0,0 +1,36 @@
+require 'flow_test_helper'
+
+class SessionsFlowTest < FlowTestCase
+  test 'log in and log out' do
+    index_page = PageObjects::Images::IndexPage.visit
+    assert index_page.showing_signup_link?
+    assert index_page.showing_login_link?
+    refute index_page.showing_logout_link?
+
+    @user = User.new(email: 'abc@gmail.com', password: 'password', password_confirmation: 'password')
```
Comment (grahamotte) [inline test/flows/sessions_flow_test.rb]: What happens if this fails?

File: app/controllers/sessions_controller.rb [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+      log_out
+    else
+      flash[:error] = 'Not logged in!'
+    end
+    redirect_to images_path
+  end
+
+  private
+
+  def check_login
```
Comment (grahamotte) [inline app/controllers/sessions_controller.rb]: It's not clear from the name what this method does

File: app/controllers/sessions_controller.rb (hunk 15 above)
Comment (RitamBanerjee) [reply app/controllers/sessions_controller.rb]: rubocop complains when I use and
